{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cc0229b6-ee3c-4881-b721-7e341338a350",
   "metadata": {},
   "source": [
    "# Import Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "da52c472-78cd-4cde-85a7-77d9e6474228",
   "metadata": {},
   "outputs": [],
   "source": [
    "# built-in\n",
    "import json\n",
    "from pathlib import Path\n",
    "\n",
    "# third-party (pip install required)\n",
    "import networkx as nx\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6006c026-ba1c-43a5-ab1a-e78c013c06ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_PATH = Path('../../data/')\n",
    "ANNOTATION_PATH = Path(DATA_PATH, 'annotation')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9f89871-68ad-4472-af3e-d3f27e1c3389",
   "metadata": {},
   "source": [
    "# Load Existing Gene Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4d37a26c-3262-4dba-9160-243055666264",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>regulator</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>essential</th>\n",
       "      <th>...</th>\n",
       "      <th>y_ome</th>\n",
       "      <th>k_eff</th>\n",
       "      <th>schmidt_prot_med</th>\n",
       "      <th>heckmann_prot_med</th>\n",
       "      <th>proteomics</th>\n",
       "      <th>p1k</th>\n",
       "      <th>p1k_ctrl_log_tpm</th>\n",
       "      <th>p1k_median_log_tpm</th>\n",
       "      <th>p1k_mad_log_tpm</th>\n",
       "      <th>in_iM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0001</th>\n",
       "      <td>thrL</td>\n",
       "      <td>thrL</td>\n",
       "      <td>thr operon leader peptide</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0AD86</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>10.372833</td>\n",
       "      <td>8.862343</td>\n",
       "      <td>1.936952</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>thrA</td>\n",
       "      <td>thrA;thrA1;thrA2</td>\n",
       "      <td>fused aspartate kinase/homoserine dehydrogenase 1</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>RpoD,Ile-tRNA,Thr-tRNA,DksA,Fnr,ArcA,ppGpp</td>\n",
       "      <td>P00561</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>4101.5</td>\n",
       "      <td>4434.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>10.314562</td>\n",
       "      <td>9.605329</td>\n",
       "      <td>1.042143</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>thrB</td>\n",
       "      <td>thrB</td>\n",
       "      <td>homoserine kinase</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>RpoD,Ile-tRNA,Thr-tRNA,DksA,ppGpp</td>\n",
       "      <td>P00547</td>\n",
       "      <td>2801</td>\n",
       "      <td>3733</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>442.0</td>\n",
       "      <td>1177.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>9.982251</td>\n",
       "      <td>9.232803</td>\n",
       "      <td>1.035654</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>thrC</td>\n",
       "      <td>thrC</td>\n",
       "      <td>threonine synthase</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>RpoD,ppGpp,Ile-tRNA,Thr-tRNA,DksA</td>\n",
       "      <td>P00934</td>\n",
       "      <td>3734</td>\n",
       "      <td>5020</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>8818.5</td>\n",
       "      <td>3516.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>9.973666</td>\n",
       "      <td>9.274271</td>\n",
       "      <td>0.906287</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>yaaX</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>DUF2502 domain-containing protein YaaX</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>Lrp</td>\n",
       "      <td>P75616</td>\n",
       "      <td>5234</td>\n",
       "      <td>5530</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>7.132482</td>\n",
       "      <td>6.283661</td>\n",
       "      <td>1.093146</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_name          synonyms  \\\n",
       "locus_tag                               \n",
       "b0001          thrL              thrL   \n",
       "b0002          thrA  thrA;thrA1;thrA2   \n",
       "b0003          thrB              thrB   \n",
       "b0004          thrC              thrC   \n",
       "b0005          yaaX              yaaX   \n",
       "\n",
       "                                                gene_product  \\\n",
       "locus_tag                                                      \n",
       "b0001                              thr operon leader peptide   \n",
       "b0002      fused aspartate kinase/homoserine dehydrogenase 1   \n",
       "b0003                                      homoserine kinase   \n",
       "b0004                                     threonine synthase   \n",
       "b0005                 DUF2502 domain-containing protein YaaX   \n",
       "\n",
       "                                           COG  \\\n",
       "locus_tag                                        \n",
       "b0001                        No COG annotation   \n",
       "b0002      Amino acid transport and metabolism   \n",
       "b0003      Nucleotide transport and metabolism   \n",
       "b0004      Amino acid transport and metabolism   \n",
       "b0005                         Function unknown   \n",
       "\n",
       "                                            regulator uniprot  start   end  \\\n",
       "locus_tag                                                                    \n",
       "b0001                                             NaN  P0AD86    190   255   \n",
       "b0002      RpoD,Ile-tRNA,Thr-tRNA,DksA,Fnr,ArcA,ppGpp  P00561    337  2799   \n",
       "b0003               RpoD,Ile-tRNA,Thr-tRNA,DksA,ppGpp  P00547   2801  3733   \n",
       "b0004               RpoD,ppGpp,Ile-tRNA,Thr-tRNA,DksA  P00934   3734  5020   \n",
       "b0005                                             Lrp  P75616   5234  5530   \n",
       "\n",
       "          strand  essential  ...  y_ome  k_eff  schmidt_prot_med  \\\n",
       "locus_tag                    ...                                   \n",
       "b0001          +      False  ...  False  False               NaN   \n",
       "b0002          +      False  ...  False   True            4101.5   \n",
       "b0003          +      False  ...  False   True             442.0   \n",
       "b0004          +      False  ...  False   True            8818.5   \n",
       "b0005          +      False  ...   True  False               NaN   \n",
       "\n",
       "           heckmann_prot_med  proteomics    p1k  p1k_ctrl_log_tpm  \\\n",
       "locus_tag                                                           \n",
       "b0001                    NaN       False  False         10.372833   \n",
       "b0002                 4434.0        True   True         10.314562   \n",
       "b0003                 1177.0        True   True          9.982251   \n",
       "b0004                 3516.0        True   True          9.973666   \n",
       "b0005                    NaN       False   True          7.132482   \n",
       "\n",
       "           p1k_median_log_tpm  p1k_mad_log_tpm  in_iM  \n",
       "locus_tag                                              \n",
       "b0001                8.862343         1.936952  False  \n",
       "b0002                9.605329         1.042143   True  \n",
       "b0003                9.232803         1.035654   True  \n",
       "b0004                9.274271         0.906287   True  \n",
       "b0005                6.283661         1.093146   True  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table = pd.read_csv(Path(ANNOTATION_PATH, 'gene_info.csv'), index_col=0)\n",
    "gene_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39f92f1f-f0c8-4835-9a61-7af5afc2bb84",
   "metadata": {},
   "source": [
    "# KEGG Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76e0e441-e42a-4caa-8fd0-0e18e783c486",
   "metadata": {},
   "source": [
    "## Load the KEGG Orthology into a Graph\n",
    "\n",
    "KEGG annotations downloaded from: https://www.genome.jp/brite/eco00001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "f28747f4-8ed9-4b38-8975-a1e5db0a9cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(Path(ANNOTATION_PATH, 'kegg_map.json'), 'r') as kegg_file:\n",
    "    kegg_map = json.load(kegg_file)\n",
    "    \n",
    "excludes = ['09180 Brite Hierarchies', '09181 Protein families: metabolism', '09182 Protein families: genetic information processing']\n",
    "    \n",
    "def get_edgelist(kegg_dict):\n",
    "    edgelist = []\n",
    "    if kegg_dict['name'] in excludes:\n",
    "        return []\n",
    "    if 'children' in kegg_dict:\n",
    "        for child_dict in kegg_dict['children']:\n",
    "            edgelist.append((kegg_dict['name'], child_dict['name']))\n",
    "            child_edgelist = get_edgelist(child_dict) \n",
    "            edgelist.extend(child_edgelist)\n",
    "    return edgelist\n",
    "\n",
    "kegg_edgelist = get_edgelist(kegg_map)\n",
    "\n",
    "kegg_graph = nx.DiGraph(kegg_edgelist)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f16b21f3-7c40-45fe-9984-b0e6cf8c7363",
   "metadata": {},
   "source": [
    "## Graph Curation\n",
    "\n",
    "KEGG allows for multiple categorizations, but for our purposes, we want to flatten this to a tree structure so that each gene has an unambiguous path from a top-level category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "34d8432b-fd1e-4b43-af62-3861d0edc148",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [gene for gene, out_deg in kegg_graph.out_degree() if out_deg == 0 and gene[0] == 'b']\n",
    "\n",
    "# there is only ambiguity if a gene has more than 4 ancestors (if perfect path to root, will have 3 categories and root noe)\n",
    "ambi_genes = [g for g in genes if len(nx.ancestors(kegg_graph, g)) > 4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4c0d7aa-65ed-455c-a60d-7211640bd999",
   "metadata": {},
   "outputs": [],
   "source": [
    "ambi_cat = {\n",
    "    'b2388': '00010',\n",
    "    'b4025': '00010',\n",
    "    'b3916': '00010',\n",
    "    'b1723': '00010',\n",
    "    'b4232': '00010',\n",
    "    'b3925': '00010',\n",
    "    'b2930': '00010',\n",
    "    'b2097': '00010',\n",
    "    'b2925': '00010',\n",
    "    'b3919': '00010',\n",
    "    'b0755': '00010',\n",
    "    'b4395': '00010',\n",
    "    'b3612': '00010',\n",
    "    'b2779': '00010',\n",
    "    'b1676': '00010',\n",
    "    'b1854': '00010',\n",
    "    'b1702': '00010',\n",
    "    'b0114': '00620',\n",
    "    'b0115': '00620',\n",
    "    'b0116': '00620',\n",
    "    'b1378': '00620',\n",
    "    'b0356': '00071',\n",
    "    'b1241': '00650',\n",
    "    'b1478': '00071',\n",
    "    'b3589': '00350',\n",
    "    'b4269': '00620',\n",
    "    'b0325': '00620',\n",
    "    'b2453': '00620',\n",
    "    'b3588': '00620',\n",
    "    'b4069': '00620',\n",
    "    'b0756': '00052',\n",
    "    'b3879': '00052',\n",
    "    'b0688': '00500',\n",
    "    'b3403': '00010',\n",
    "    'b1734': '00500',\n",
    "    'b2716': '00500',\n",
    "    'b2901': '00500',\n",
    "    'b3721': '00500',\n",
    "    'b2417': '02060',\n",
    "    'b1101': '02060',\n",
    "    'b1621': '02060',\n",
    "    'b2715': '02060',\n",
    "    'b0720': '00020',\n",
    "    'b1276': '00020',\n",
    "    'b0771': '00020',\n",
    "    'b0118': '00020',\n",
    "    'b1136': '00020',\n",
    "    'b0727': '00020',\n",
    "    'b0729': '00020',\n",
    "    'b0728': '00020',\n",
    "    'b0723': '00020',\n",
    "    'b0724': '00020',\n",
    "    'b0721': '00020',\n",
    "    'b0722': '00020',\n",
    "    'b4154': '00190',\n",
    "    'b4153': '00190',\n",
    "    'b4152': '00190',\n",
    "    'b4151': '00190',\n",
    "    'b4122': '00020',\n",
    "    'b1612': '00020',\n",
    "    'b1611': '00020',\n",
    "    'b1675': '00020',\n",
    "    'b2929': '00020',\n",
    "    'b3236': '00020',\n",
    "    'b2210': '00020',\n",
    "    'b1852': '00030',\n",
    "    'b2029': '00030',\n",
    "    'b3386': '00030',\n",
    "    'b2465': '00030',\n",
    "    'b2935': '00030',\n",
    "    'b4090': '00030',\n",
    "    'b4383': '00030',\n",
    "    'b3380': '00030',\n",
    "    'b1207': '00030',\n",
    "    'b1850': '00030',\n",
    "    'b3553': '00630',\n",
    "    'b0772': '00040',\n",
    "    'b1617': '00040',\n",
    "    'b2028': '00541',\n",
    "    'b1236': '00541',\n",
    "    'b2042': '00541',\n",
    "    'b3575': '00053',\n",
    "    'b3580': '00053',\n",
    "    'b4196': '00053',\n",
    "    'b3581': '00053',\n",
    "    'b4197': '00053',\n",
    "    'b3582': '00053',\n",
    "    'b4198': '00053',\n",
    "    'b0061': '00040',\n",
    "    'b3583': '00053',\n",
    "    'b3565': '00040',\n",
    "    'b3904': '00040',\n",
    "    'b3902': '00040',\n",
    "    'b1774': '00040',\n",
    "    'b2802': '00040',\n",
    "    'b2803': '00040',\n",
    "    'b2800': '00040',\n",
    "    'b0394': '00051',\n",
    "    'b1613': '00051',\n",
    "    'b2048': '00051',\n",
    "    'b2049': '00051',\n",
    "    'b2053': '00051',\n",
    "    'b2052': '00541',\n",
    "    'b2168': '00051',\n",
    "    'b2167': '02060',\n",
    "    'b1817': '02060',\n",
    "    'b1818': '02060',\n",
    "    'b1819': '02060',\n",
    "    'b2934': '02060',\n",
    "    'b2933': '02060',\n",
    "    'b3599': '02060',\n",
    "    'b2704': '02060',\n",
    "    'b2703': '02060',\n",
    "    'b2702': '02060',\n",
    "    'b0757': '00052',\n",
    "    'b0758': '00052',\n",
    "    'b0759': '00541',\n",
    "    'b0344': '00052',\n",
    "    'b3076': '00052',\n",
    "    'b3077': '00052',\n",
    "    'b2036': '00541',\n",
    "    'b4119': '00052',\n",
    "    'b4302': '02060',\n",
    "    'b2094': '02060',\n",
    "    'b2093': '02060',\n",
    "    'b4565': '02060',\n",
    "    'b4304': '02060',\n",
    "    'b0403': '00500',\n",
    "    'b3133': '02060',\n",
    "    'b3138': '02060',\n",
    "    'b3139': '02060',\n",
    "    'b3140': '02060',\n",
    "    'b4195': '02060',\n",
    "    'b4194': '02060',\n",
    "    'b4193': '02060',\n",
    "    'b2132': '00500',\n",
    "    'b3533': '02026',\n",
    "    'b1736': '02060',\n",
    "    'b1738': '02060',\n",
    "    'b1737': '02060',\n",
    "    'b3430': '00500',\n",
    "    'b3429': '00500',\n",
    "    'b3417': '00500',\n",
    "    'b3428': '00500',\n",
    "    'b3682': '02060',\n",
    "    'b3683': '02060',\n",
    "    'b4240': '02060',\n",
    "    'b1107': '00520',\n",
    "    '02060': '00520',\n",
    "    'b3729': '00520',\n",
    "    'b3730': '09107',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "1d9dc347-b639-4ceb-88c3-8de4ccb4e7ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b3730 glmU; fused N-acetylglucosamine-1-phosphate uridyltransferase and glucosamine-1-phosphate acetyltransferase\tK04042 glmU; bifunctional UDP-N-acetylglucosamine pyrophosphorylase / glucosamine-1-phosphate N-acetyltransferase [EC:2.7.7.23 2.3.1.157]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'00520 Amino sugar and nucleotide sugar metabolism [PATH:eco00520]',\n",
       " '00541 O-Antigen nucleotide sugar biosynthesis [PATH:eco00541]',\n",
       " '09100 Metabolism',\n",
       " '09101 Carbohydrate metabolism',\n",
       " '09107 Glycan biosynthesis and metabolism',\n",
       " 'eco00001'}"
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_to_test = ambi_genes[150]\n",
    "print(gene_to_test)\n",
    "\n",
    "nx.ancestors(kegg_graph, gene_to_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "0849f1d4-8dda-47de-b5e2-179079cb5e28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'09100 Metabolism', 'eco00001'}"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nx.ancestors(kegg_graph, '09105 Amino acid metabolism')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98058d4b-12da-4432-b740-68f0ce2f34d1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
